src/SocketTrebuchet.ts

```diff
@@ -45,18 +49,53 @@ class SocketTrebuchet extends Trebuchet {
     }, this.timeout * 1.5)
   }
 
+  private respondToReliableMessage (decodedData: any) {
+    const synId = decodedData.synId
+    const attempt = decodedData.attempt
+    const ackId = (synId << 2) | attempt
+    this.ws.send(new Uint8Array([ACK_PREFIX, ackId]))
+    this.lastReliableSynId = synId === MAX_SYN_ID - 1 ? -1 : synId
+    this.emit('data', decodedData.object)
+  }
+
+  private processReliableMessageInOrder (decodedData: any) {
+    const synId = decodedData.synId
+    this.reliableMessageQueue
+      .filter((_data, idx) => idx < synId)
+      .forEach((data, idx) => {
+        this.respondToReliableMessage(data)
+        delete this.reliableMessageQueue[idx]
+      })
+    this.respondToReliableMessage(decodedData)
+  }
+
   protected setup () {
     this.ws = new WebSocket(this.getUrl(), TREBUCHET_WS)
     this.ws.binaryType = 'arraybuffer'
     this.ws.onopen = this.handleOpen.bind(this)
+    this.lastReliableSynId = -1
+    this.reliableMessageQueue = []
 
     this.ws.onmessage = (event: MessageEvent) => {
       const {data} = event
       if (isPing(data)) {
         this.keepAlive()
         this.ws.send(PONG)
       } else {
-        this.emit('data', this.decode(data))
+        const decodedData = this.decode(data)
+        const synId = decodedData.synId
+        if (synId !== undefined) {
+          if (
+            this.lastReliableSynId + 1 === synId ||
+            (this.lastReliableSynId + 1 === MAX_SYN_ID && synId === 0)
```

Comment:
There is no bug but rather a redundant condition check. According to L57, after we process `synId` 63, we set `lastReliableSynId` to -1. Next time we are expecting `synId` 0 so it will just satisfy the first part of the OR condition: `this.lastReliableSynId + 1 === synId`.
Anyway, I can clean this code to make it easier to read & debug.